cody: toggle config for inline and non-stop no reload required

abeatrix opened the pull request:
Requested by https://github.com/sourcegraph/sourcegraph/pull/53307
Close https://github.com/sourcegraph/sourcegraph/issues/53351 https://github.com/sourcegraph/sourcegraph/issues/53363

This PR includes:
- Updating the configuration for Inline Assist no longer requires users to reload the extension.
- Remove Code Lens to save Inline Fixup
- Add Undo Code Lens to Inline Fixup jobs that undo changes made by Cody
- Remove all comment threads from file when file is closed

For details: https://www.loom.com/share/540f4953b1ac4c9d81e90f4c37f86c2d?sid=41c65fd7-1be2-4225-92eb-3c7fa56766f8

## Test plan

<!-- All pull requests REQUIRE a test plan: https://docs.sourcegraph.com/dev/background-information/testing_principles -->

All the existing tests should be passing. 


https://github.com/sourcegraph/sourcegraph/assets/68532117/ad229361-23ba-4653-91f5-b9dae98ee765

abeatrix commented on client/cody/src/services/CodeLensProvider.ts:
```diff
@@ -57,13 +53,29 @@ export class CodeLensProvider implements vscode.CodeLensProvider {
+        this.contextStore.set(id, { docUri, original, replacement })
+    }
+
+    public async undo(id: string): Promise<void> {
```
Clicking on undo will replace the replacement context with the original context

dominiccooney commented on client/cody/CHANGELOG.md:
```diff
@@ -8,8 +8,10 @@ Starting from `0.2.0`, Cody is using `major.EVEN_NUMBER.patch` for release versi
+- Add delete button for removing individual history. [pull/52904](https://github.com/sourcegraph/sourcegraph/pull/52904)
+- Load the recent ongoing chat on reload of window. [pull/52904](https://github.com/sourcegraph/sourcegraph/pull/52904)
+- Inline Assist: New Code Lens to undo `inline fix` performed by Cody. [pull/53348](https://github.com/sourcegraph/sourcegraph/pull/53348)
+- Inline Assist: New Code Lens to revert file back to last committed state. [pull/53348](https://github.com/sourcegraph/sourcegraph/pull/53348)
```
Adding a way to revert to a saved file in the code lens doesn't make sense to me. Reverting the file has effects way outside what the code lens relates to. If the user wants to revert the file, they should use File > Revert File or git > Revert Selected Ranges or something like that. Let's not duplicate this general functionality into our code lens.

dominiccooney commented on client/cody/src/editor/vscode-editor.ts:
```diff
@@ -19,7 +19,33 @@ export class VSCodeEditor implements Editor {
+            const isTesting = process.env.CODY_TESTING === 'true'
+            if (e.affectsConfiguration('cody')) {
+                // Non-Stop Cody
+                const enableNonStop = (config.get('experimental.nonStop') as boolean) || isTesting
```
Let's not make non-stop fixups configurable at runtime. There's work to cleanly shut down things in flight if you can toggle this dynamically that isn't being attempted here. So please back this part of the change out.

Inline assist, I leave to you judgement, but you should test cases like having an inline assist spinning and then turning the feature off; having it on, spin, off, on again, etc.

On the whole I don't think this is worth the complexity it might entail. Do people want to flip the availability of these features off and on a lot?

abeatrix replied:
Yea I will remove Non-Stop Cody from this. 

> Do people want to flip the availability of these features off and on a lot?

 I think we just want to make it easier so that users do need to reload their IDE when they enable the feature.

dominiccooney reviewed (changes requested):
- We shouldn't add a "revert" to the code lens, it seems unrelated. The lenses have a scope within the file; revert applies over the whole file.
- Don't make non-stop fixup dynamic in this way. This needs care to handle in-flight work work shutting down cleanly. I don't think it's worth it because it will complicate respinning, etc. but if we want to do it, it needs tests and probably more work.
- Up to you about inline chat but do similar concerns apply?

abeatrix commented:
@dominiccooney updated based on your last review

- We shouldn't add a "revert" to the code lens, it seems unrelated. The lenses have a scope within the file; revert applies over the whole file.

  - Removed.

- Don't make non-stop fixup dynamic in this way. This needs care to handle in-flight work work shutting down cleanly. I don't think it's worth it because it will complicate respinning, etc. but if we want to do it, it needs tests and probably more work.

  - Removed toggle for Non-Stop fixup

- Up to you about inline chat but do similar concerns apply?

  - Keeping this for Inline Chat because the toggle is for the ability to create a comment thread. It does not affect any inflight fixup job is handled by the Fixup recipe.

abeatrix commented:
@dominiccooney updated based on your last review

- We shouldn't add a "revert" to the code lens, it seems unrelated. The lenses have a scope within the file; revert applies over the whole file.

  - Removed.

- Don't make non-stop fixup dynamic in this way. This needs care to handle in-flight work work shutting down cleanly. I don't think it's worth it because it will complicate respinning, etc. but if we want to do it, it needs tests and probably more work.

  - Removed toggle for Non-Stop fixup

- Up to you about inline chat but do similar concerns apply?

  - Keeping this for Inline Chat because the toggle is for the ability to create a comment thread. It does not affect any inflight fixup job is handled by the Fixup recipe.

dominiccooney reviewed (approved):
Looks good.

Would be great to spell out the command to run the existing tests in the test plan. It will help others grok how to run those tests.
